<hr size="4" color="blue">
<p style="text-align: center;"><span style="color: #0000ff;"><font size="+5"><strong> 3 Affichage</strong> </font></span> </p>
<hr size="4" color="blue">

#  <span style="color: #0000ff;"><font size="+2"><strong>3.1 Sortie écran </strong> </font> </span>

<div style="border: 2px solid #C0392B; padding: 3px; background-color: #E6B0AA; -moz-border-radius-topleft: 5px; -moz-border-radius-topright: 5px; -moz-border-radius-bottomright: 5px; -moz-border-radius-bottomleft: 5px;">
    

**La fonction <span style="background:#E7E7E7">*print()*</span> affiche l'argument qu'on lui passe entre parenthèses et un retour à ligne.**
Ce retour à ligne supplémentaire est ajouté par défaut. Si toutefois, on ne veut pas afficher ce retour à la ligne, 
on peut utiliser l'argument par « mot-clé » end :

On utilise l'instruction ***print()*** classiquement en passant la chaîne de caractères "Hello world!" en argument.

In [ ]:
print("Hello world!")

On a ajouté un second argument end="", en précisant le mot-clé end qui modifie le comportement par défaut des fonctions.

In [2]:
print("Hello world!", end="a")

Hello world!a

Utiliser deux fonctions print() à la suite. Dans la portion de code suivante, le caractère « ; » sert à séparer plusieurs instructions Python sur une même ligne :

In [4]:
print("Hello");  print("Joe")

Hello
Joe


In [5]:
print("Hello", end="a") ; print("Joe")

HelloaJoe


La fonction print() peut également ***afficher le contenu d'une variable*** quel que soit son type. 
Par exemple, pour un entier :

In [7]:
var = 3
print(var)

3


Il est également possible d'afficher le contenu de plusieurs variables (quel que soit leur type)
en les séparant par des virgules :

In [8]:
x = 32
nom = "John"
print(nom, "a", x, "ans")

John a 32 ans


Python a écrit une phrase complète en remplaçant les variables x et nom par leur contenu. Vous remarquerez que pour afficher plusieurs éléments de texte sur une seule ligne, nous avons utilisé le séparateur « , » entre les différents éléments. Python a également ajouté un espace à chaque fois que l'on utilisait le séparateur « , ». On peut modifier ce comportement en passant à la fonction print() l'argument par mot-clé sep :

In [11]:
x = 32
nom = "John"
print(nom, "a", x, "ans", sep="")

Johna32ans


In [12]:
print(nom, "a", x, "ans", sep="-")

John-a-32-ans


Pour afficher deux chaînes de caractères l'une à côté de l'autre, sans espace, on peut soit les concaténer, 
soit utiliser l'argument par mot-clé sep avec une chaîne de caractères vide :

In [13]:
ani1 = "chat"
ani2 = "souris"
print(ani1, ani2)

chat souris


In [14]:
print(ani1 + ani2)
print(ani1, ani2, sep="çç")

chatsouris
chatççsouris


 #  <span style="color: #0000ff;"><font size="+2"><strong>3.2 Écriture formatée</strong> </font> </span>

La méthode <span style="background:#E7E7E7">*.format()*</span> permet une meilleure organisation de l'affichage des variables

Si on reprend l'exemple précédent :

In [15]:
x = 32
nom = "John"
print("{} a {} ans".format(nom, x))

John a 32 ans


Dans la chaîne de caractères,
- les accolades vides {} précisent l'endroit où le contenu de la variable doit être inséré.
- Juste après la chaîne de caractères, l'instruction .format(nom, x) fournie la liste des variables à insérer, d'abord la variable nom puis la variable x. <br>

La méthode .format() agit sur la chaîne de caractères à laquelle elle est attachée par le point.

 ***Remarque***
 
Il est possible d'indiquer entre les accolades {} dans quel ordre afficher les variables, avec 0 pour la variable à afficher en premier, 1 pour la variable à afficher en second, etc. (attention, Python commence à compter à 0). Cela permet de modifier l'ordre dans lequel sont affichées les variables.

In [18]:
x = 32
nom = "John"

In [19]:
print("{1} a {0} ans".format(nom, x))

32 a John ans


Soit la probabilité suivante

In [20]:
prop= (4500 + 2575) / 14800
print("La proportion de GC est", prop)

La proportion de GC est 0.4780405405405405


                                   
Le résultat obtenu présente trop de décimales (seize dans le cas présent).
Pour écrire le résultat plus lisiblement, vous pouvez spécifier dans les accolades {} le format qui vous intéresse. Dans le cas présent, vous voulez formater un float pour l'afficher avec deux puis trois décimales :

In [21]:
print("La proportion de GC est {:.2f}".format(prop))

La proportion de GC est 0.48


Détaillons le contenu des accolades de la première ligne ({:.2f}) :

- Les deux points : indiquent qu'on veut préciser le format.
- la lettre f indique qu'on souhaite afficher la variable sous forme d'un float.
- Les caractères .2 indiquent la précision voulue, soit ici deux chiffres après la virgule.
- Notez enfin que le formatage avec .xf (x étant un entier positif) renvoie un résultat arrondi.

Il est par ailleurs possible de combiner le formatage (à droite des 2 points) ainsi que l'emplacement des variables à substituer (à gauche des 2 points), par exemple :

In [26]:
print("prop GC(2 déci.) = {0:.2f}, prop GC(3 déci.) = {0:.3f}".format(prop))

prop GC(2 déci.) = 0.48, prop GC(3 déci.) = 0.478


Vous remarquerez qu'on utilise ici la même variable (prop_GC) à deux endroits différents.

Vous pouvez aussi formater des entiers avec la lettre d :

In [23]:
nb_G = 4500
print("Ce génome contient {:d} guanines".format(nb_G))

Ce génome contient 4500 guanines


ou mettre plusieurs nombres dans une même chaîne de caractères.

In [27]:
nb_G = 4500
nb_C = 2575
print("Ce génome contient {:d} G et {:d} C, soit une prop de GC de {:.2f}".format(nb_G,nb_C,prop))

Ce génome contient 4500 G et 2575 C, soit une prop de GC de 0.48


In [25]:
perc_GC = prop_GC * 100

NameError: name 'prop_GC' is not defined

---
<p style="text-align: center;"><font size="+2"><strong>Pour aller plus loin</strong> </font> </p>

---


>Enfin, il est possible de préciser sur combien de caractères vous voulez qu'un résultat soit écrit et comment se fait l'alignement (à gauche, à droite ou centré). Dans la portion de code suivante, le caractère ; sert de séparateur entre les instructions sur une même ligne :

In [7]:
print(10) ; print(1000)

10
1000


In [8]:
print("{:>6d}".format(10)) ; print("{:>6d}".format(1000))
print("{:<6d}".format(10)) ; print("{:<6d}".format(1000))
print("{:^6d}".format(10)) ; print("{:^6d}".format(1000))
print("{:*^6d}".format(10)) ; print("{:*^6d}".format(1000))
print("{:0>6d}".format(10)) ; print("{:0>6d}".format(1000))

    10
  1000
10    
1000  
  10  
 1000 
**10**
*1000*
000010
001000


> Notez que > spécifie un alignement à droite, < spécifie un alignement à gauche et ^ spécifie un alignement centré. Il est également possible d'indiquer le caractère qui servira de remplissage lors des alignements (l'espace est le caractère par défaut).
Ce formatage est également possible sur des chaînes de caractères avec la lettre s (comme string) :

In [9]:
print("atom HN") ; print("atom HDE1")

atom HN
atom HDE1


>Vous voyez tout de suite l'énorme avantage de l'écriture formatée.
Elle vous permet d'écrire en colonnes parfaitement alignées. Pour les floats, il est possible de combiner le nombre de caractères à afficher avec le nombre de décimales :

In [17]:
print("{:7.3f}".format(prop))
print("{:10.3f}".format(prop))

  0.478
     0.478


> L'instruction 7.3f signifie que l'on souhaite écrire un float avec 3 décimales et formaté sur 7 caractères
(par défaut justifiés à droite). L'instruction 10.3f fait la même chose sur 10 caractères.
Remarquez que le séparateur décimal . compte pour un caractère.

> Enfin, si on veut afficher des accolades littérales et utiliser la méthode .format() en même temps, 
il faut doubler les accolades pour échapper au formatage :

    

In [15]:
print("Accolades littérales {{}} et pour le formatage {}".format(10))

Accolades littérales {} et pour le formatage 10


> Accolades littérales {} et pour le formatage 10
La méthode .format() agit sur la chaîne de caractères à laquelle elle est attachée par un point
et n'a rien à voir avec la fonction print(). Si on donne une chaîne de caractères suivie d'un .format()
à la fonction print(), Python évalue d'abord le formatage et c'est la chaîne de caractères
qui en résulte qui est affichée à l'écran. Tout comme dans l'instruction print(5*5), 
c'est d'abord la multiplication (5*5) qui est évaluée, puis son résultat qui est affiché à l'écran. 
On peut s'en rendre compte de la manière suivante dans l'interpréteur :

In [18]:
"{:10.3f}".format(prop)

'     0.478'

In [20]:
 type("{:10.3f}".format(prop))

str

> Python affiche le résultat de l'instruction "{:10.3f}".format(perc_GC) 
comme une chaîne de caractères et la fonction type() nous le confirme.

  
<div style="border: 2px solid #0000ff; padding: 3px; background-color: #0099FF; -moz-border-radius-topleft: 5px; -moz-border-radius-topright: 5px; -moz-border-radius-bottomright: 5px; -moz-border-radius-bottomleft: 5px;">

***Remarque***

Revenons quelques instants sur la notion de méthode abordée dans ce chapitre avec .format(). En Python, on peut considérer chaque variable comme un objet sur lequel on peut appliquer des méthodes. Une méthode est simplement une fonction qui utilise et/ou agit sur l'objet lui-même, les deux étant connectés par un point. La syntaxe générale est de la forme objet.méthode().

Par exemple dans,

"Joe a {} ans".format(20)

la méthode .format() est liée à "Joe a {} ans" qui est un objet de type chaîne de caractères. La méthode renvoie une nouvelle chaîne de caractères avec le bon formatage (ici, 'Joe a 20 ans').

--------------
<p style="text-align: center;"><span style="color: #0000ff;"><font size="+3"><strong>Exercices</strong> </font> </span></p>

------------------


### 3.5.4 Écriture formatée
En utilisant l'écriture formatée, affichez en une seule ligne les variables a, b et c dont les valeurs sont respectivement la chaîne de caractères "salut", le nombre entier 102 et le float 10.318. La variable c sera affichée avec 2 décimales.

### 3.5.5 Écriture formatée 
- Calculez un pourcentage de GC avec l'instruction suivante :

perc_GC = ((4500 + 2575)/14800)*100

- affichez le contenu de la variable perc_GC à l'écran avec 0, 1, 2 puis 3 décimales sous forme arrondie en utilisant .format(). On souhaite que le programme affiche la sortie suivante :


Le pourcentage de GC est 48     %  
Le pourcentage de GC est 47.8   %  
Le pourcentage de GC est 47.80  %  
Le pourcentage de GC est 47.804 %